launch_ros/launch_ros/descriptions/composable_node.py

```diff
@@ -0,0 +1,113 @@
+        package: SomeSubstitutionsType,
+        node_plugin: SomeSubstitutionsType,
+        node_name: Optional[SomeSubstitutionsType] = None,
+        node_namespace: Optional[SomeSubstitutionsType] = None,
+        parameters: Optional[SomeParameters] = None,
+        remappings: Optional[SomeRemapRules] = None,
+        extra_arguments: Optional[SomeParameters] = None,
+    ) -> None:
+        """
+        Initialize a ComposableNode description.
+
+        :param package: name of the ROS package the node plugin lives in
+        :param node_plugin: name of the plugin to be loaded
+        :param node_name: name the node should have
+        :param node_namespace: namespace the node should create topics/services/etc in
+        :param parameters: list of either paths to yaml files or dictionaries of parameters
+        :param remappings: list of from/to pairs for remapping names
+        :pram extra_arguments: container specific arguments to be passed to the loaded node
+        """
+        self.__package = normalize_to_list_of_substitutions(package)
+        self.__node_plugin = normalize_to_list_of_substitutions(node_plugin)
+
+        self.__node_name = None  # type: Optional[List[Substitution]]
+        if node_name is not None:
+            self.__node_name = normalize_to_list_of_substitutions(node_name)
+
+        self.__node_namespace = None  # type: Optional[List[Substitution]]
+        if node_namespace is not None:
+            self.__node_namespace = normalize_to_list_of_substitutions(node_namespace)
+
+        self.__parameters = None  # type: Optional[Parameters]
+        if parameters is not None:
+            self.__parameters = normalize_parameters(parameters)
+
+        self.__remappings = None  # type: Optional[RemapRules]
+        if remappings:
+            self.__remappings = normalize_remap_rules(remappings)
+
+        self.__extra_arguments = None  # type: Optional[Parameters]
+        if extra_arguments:
+            self.__extra_arguments = normalize_parameters(extra_arguments)
+
+    @property
+    def package(self) -> List[Substitution]:
+        """Get node package name as a sequence of substitutions to be performed."""
+        return self.__package
+
+    @property
+    def node_plugin(self) -> List[Substitution]:
+        """Get node plugin name as a sequence of substitutions to be performed."""
+        return self.__node_plugin
+
+    @property
+    def node_name(self) -> Optional[List[Substitution]]:
+        """Get node name as a sequence of substitutions to be performed."""
+        return self.__node_name
+
+    @property
+    def node_namespace(self) -> Optional[List[Substitution]]:
+        """Get node namespace as a sequence of substitutions to be performed."""
+        return self.__node_namespace
+
+    @property
+    def parameters(self) -> Optional[Parameters]:
+        """Get node parameter YAML files or dicts with substitutions to be performed."""
+        return self.__parameters
```

Comment:
Since everything is a reference and this can contain dictionaries which are mutable, I had been thinking about the description doing a `copy.deepcopy` to guard against a buggy action modifying a description that was used multiple places. I'm don't think it needs to be done, but thought I'd mention it in case that makes someone think of something.

Reply:
+1 to keep descriptions immutable, though the same could apply to actions in general and we don't try to enforce it elsewhere.

Reply:
Well, we do try by using properties, but that's as far as we go.